Userland/Utilities/pkg/AvailablePort.cpp

```diff
@@ -0,0 +1,176 @@
+    };
+    request->stream_into(*output_stream);
+    return loop.exec();
+}
+
+ErrorOr<HashMap<String, AvailablePort>> AvailablePort::read_available_ports_list()
+{
+    auto available_ports_file = TRY(Core::File::open("/usr/Ports/AvailablePorts.md"sv, Core::File::OpenMode::Read, 0600));
+    auto content_buffer = TRY(available_ports_file->read_until_eof());
+    auto content = StringView(content_buffer);
+    auto document = Markdown::Document::parse(content);
+    auto finder = MarkdownTableFinder::analyze(*document);
+    if (finder.table_count() != 1)
+        return Error::from_string_literal("Invalid tables count in /usr/Ports/AvailablePorts.md");
+
+    VERIFY(finder.tables()[0]);
+    auto possible_port_name_column_index = get_column_in_table(*finder.tables()[0], "Port"sv);
+    auto possible_port_version_column_index = get_column_in_table(*finder.tables()[0], "Version"sv);
+    auto possible_port_website_column_index = get_column_in_table(*finder.tables()[0], "Website"sv);
+
+    if (!possible_port_name_column_index.has_value())
+        return Error::from_string_literal("pkg: Port column not found /usr/Ports/AvailablePorts.md");
+    if (!possible_port_version_column_index.has_value())
+        return Error::from_string_literal("pkg: Version column not found /usr/Ports/AvailablePorts.md");
+    if (!possible_port_website_column_index.has_value())
+        return Error::from_string_literal("pkg: Website column not found /usr/Ports/AvailablePorts.md");
+
+    auto& port_name_column_index = possible_port_name_column_index.release_value();
+    auto& port_version_column_index = possible_port_version_column_index.release_value();
+    auto& port_website_column_index = possible_port_website_column_index.release_value();
+
+    VERIFY(port_name_column_index.rows.size() == port_version_column_index.rows.size());
+    VERIFY(port_version_column_index.rows.size() == port_website_column_index.rows.size());
+
+    HashMap<String, AvailablePort> available_ports;
+    for (size_t port_index = 0; port_index < port_name_column_index.rows.size(); port_index++) {
+        struct : public Markdown::Visitor {
+            virtual RecursionDecision visit(Markdown::Text::LinkNode const& node) override
+            {
+                text_node = node.text.ptr();
+                return RecursionDecision::Break;
+            }
+
+        public:
+            Markdown::Text::Node* text_node;
+        } text_node_find_visitor;
+
+        struct : public Markdown::Visitor {
+            virtual RecursionDecision visit(Markdown::Text::TextNode const& node) override
+            {
+                auto name_or_error = String::from_deprecated_string(node.text); // builder.to_string();
+                if (!name_or_error.is_error())
+                    name = name_or_error.release_value();
+                return RecursionDecision::Recurse;
+            }
+
+        public:
+            String name;
+        } text_node_extract_visitor;
```

Comment:
This dance you are doing with the two visitors is very confusing. 

Can we extract this out into a helper function at least? I don't think that the unnamed structs are a very common pattern in the codebase.

Why do you need to visit the text node after finding it, just to ask it its name?

Something like `ErrorOr<String> extract_port_name_from_column(Markdown::Table::Column&, size_t row_idx)` ?